compiler/il/OMRNode.cpp

```diff
@@ -2357,7 +2357,101 @@ OMR::Node::isNotCollected()
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ */
+bool
+OMR::Node::computeIsInternalPointer()
+   {
+   if (self()->getOpCode().hasPinningArrayPointer())
+      return false;
+   return self()->computeIsCollectedReference();
+   }
+
+/**
+ * \brief
+ *    Check if the node is collectable.
+ *
+ * \return
+ *    Return true if the node is collectable, false otherwise.
+ *
+ * \note
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from a collectable reference, i.e. anything that is an address to a collected object.
+ */
+bool
+OMR::Node::computeIsCollectedReference()
+   {
+   TR::Node *curNode = self();
+   if (!curNode->getType().isAddress())
+      return false;
+
+   while (curNode)
+      {
+      if (curNode->isInternalPointer())
+         return true;
 
+      if (curNode->getOpCode().hasSymbolReference())
+         {
+         TR::ILOpCode op = curNode->getOpCode();
+         TR::Symbol *symbol = curNode->getSymbolReference()->getSymbol();
+         // Current implementation treats all object references collectable, and also
+         // assumes that the return of an acall* is an object reference.
+         if (op.isNew() ||
+             (op.getDataType() == TR::Address && op.isCall()) ||
+             symbol->isInternalPointer() ||
+             symbol->isCollectedReference())
+             return true;
+         }
+
+      if (curNode->getNumChildren() >= 1)
+         {
+         TR::ILOpCode childOp = curNode->getFirstChild()->getOpCode();
+         TR::ILOpCodes ops = childOp.getOpCodeValue();
+         switch (ops)
+            {
+            case TR::aload:
+            case TR::aloadi:
+            case TR::aRegLoad:
+            case TR::areturn:
+            case TR::acall:
+            case TR::acalli:
+            case TR::compressedRefs:
+            case TR::New:
+            case TR::newarray:
+            case TR::anewarray:
+            case TR::variableNew:
+            case TR::variableNewArray:
+            case TR::multianewarray:
+            case TR::aiadd:
+            case TR::aladd:
+               curNode = curNode->getFirstChild();
+               break;
+            case TR::aternary:
+               {
+               bool isFirstChildCollected = curNode->getFirstChild()->computeIsCollectedReference();
+               bool isSecondChildCollected = curNode->getSecondChild()->computeIsCollectedReference();
+               TR_ASSERT(isFirstChildCollected != isSecondChildCollected, "Computation does not support address node from aternary with one child collectable the other uncollectable, node is " POINTER_PRINTF_FORMAT, self());
+               return isFirstChildCollected && isSecondChildCollected;
+               }
+            default:
+               return false;
+            }
+         }
+      else
```

Comment:
Without this `else` the loop would restart with the same value of `curNode`. In this case there is no child that `curNode` could have been set to. Maybe we should invert the condition to make the control flow clearer?

    if (curNode->getNumChildren() == 0)
        return false;

And then go on assuming there are children without all of the rightward drift